### DISCUSSION TITLE
Better description of redisCommandArgv()

### DISCUSSION THREAD

**Discussion by sebflaesch** [redis/hiredis#1138] [Q&A]
Hi,
Can someone write a more detailed description of redisCommandArgv() usage (or a sample)?
I don't get the idea.
Has redisCommandArgv() to be used with redisCommand() ?
Found no example in "examples".
I will dig into the source code but would be nice to have a better description in the README.
Thanks!
Seb

**Comment by michael-grunder** [comment]
I can add an example that uses `redisCommandArgv` to the basic example.

In the meantime, I created a simple function that uses it.  Essentially you are just sending two arrays with the command and arguments you wish to execute on Redis.  The first array is the arguments themselves, and the second array is the length of each argument.

In general using this command will be faster than the `redisComand` type that have to process a `printf`-alike format string.


```c
static void example_argv_command(redisContext *c, size_t n) {
    char **argv, tmp[24];
    size_t *argvlen;
    redisReply *reply;

    /* We're allocating two additional elements for command and key */
    argv = malloc(sizeof(*argv) * (2 + n));
    argvlen = malloc(sizeof(*argvlen) * (2 + n));

    /* First the command */
    argv[0] = (char*)"RPUSH";
    argvlen[0] = sizeof("RPUSH") - 1;

    /* Now our key */
    argv[1] = (char*)"argvlist";
    argvlen[1] = sizeof("argvlist") - 1;

    /* Now add the entries we wish to add to the list */
    for (size_t i = 2; i < (n + 2); i++) {
        argvlen[i] = snprintf(tmp, sizeof(tmp), "argv-element-%zu", i - 2);
        argv[i] = strdup(tmp);
    }

    /* Execute the command using redisCommandArgv.  We're sending the arguments with
     * two explicit arrays.  One for each argument's string, and the other for its
     * length. */
    reply = redisCommandArgv(c, n + 2, (const char **)argv, (const size_t*)argvlen);

    if (reply == NULL || c->err) {
        fprintf(stderr, "Error:  Couldn't execute redisCommandArgv\n");
        exit(1);
    }

    if (reply->type == REDIS_REPLY_INTEGER) {
        printf("%s reply: %lld\n", argv[0], reply->integer);
    }

    freeReplyObject(reply);

    /* Clean up */
    for (size_t i = 2; i < (n + 2); i++) {
        free(argv[i]);
    }

    free(argv);
    free(argvlen);
}

```

**Comment by sebflaesch** [comment]
Thank you Michael,

Your example is passing a set of string values for a list object...

I was wondering if it's possible to pass different types of data (number vs string), with redisCommandArgv() ...

I mean, to execute a command like:

    SET user:123 "Max Taylor" EX 60

=> Is this possible?

Background: I want to write a C API for our solution on top of hiredis (basically, it's a set of wrapper functions to hiredis API)
I need some level of abstraction, and redisCommandArgv() could help, but it must be possible to execute ANY kind of Redis command with it!

Seb

**Reply by michael-grunder** [reply]
`redisCommandArgv` only takes strings and lengths since it is a low-level function that will serialize the data into a stream of bytes for transmission over the network.

In order to do what you're thinking about, you would need to wrap the function with your own API.

```c
struct my_redis_cmd {
    char **argv;
    size_t *argvlen;
    size_t args;
};

void append_redis_arg_len(struct my_redis_cmd *cmd, const char *str, size_t len) {
    // You need other code to dynamically grow your arrays.
    // grow_arg_array();

    cmd->argv[cmd->args] = strndup(str, len);
    cmd->argvlen = len;
    cmd->args++;
}

void append_redis_arg_long(struct my_redis_cmd *cmd, long v) {
    char tmp[42];
    size_t len;
    
    // Convert the argument to a string
    len = snprintf(tmp, sizeof(tmp), "%ld", v);
    
    // now append the argument as a string.
    append_redis_arg_len(cmd, tmp, len);
}
```

It's not using hiredis but [we do this kind of thing in PhpRedis](https://github.com/phpredis/phpredis/blob/ef4699c726300a631fc9e1439b1c2aaa2004a3e6/library.c#L1016).

If you don't need to use C you could also do this much easier in a higher-level language.

**Reply by sebflaesch** [reply]
Thanks I will give it a try ASAP
We have to use C.
Seb
